src/tbb/cgroup_info.h

```diff
@@ -0,0 +1,314 @@
+            return true;
+        }
+
+        errno = 0; // Reset errno before strtoll
+        char* str_end = nullptr;
+        long long cpu_quota = std::strtoll(cpu_quota_str, &str_end, /*base*/ 10);
+        if (errno == ERANGE || str_end == cpu_quota_str)
+            return false;
+
+        num_cpus = determine_num_cpus(cpu_quota, cpu_period);
+        return num_cpus != error_value; // Return true if valid number of CPUs was determined
+    }
+
+    static int parse_cgroup_entry(const char* mnt_dir, const char* mnt_type, FILE* cgroup_fd,
+                                  cgroup_paths& paths_cache)
+    {
+        int num_cpus = error_value; // Initialize to an impossible value
+        char dir[PATH_MAX] = {0};
+        if (!std::strncmp(mnt_type, "cgroup2", 7)) { // Found cgroup v2 mount entry
+            // At first, try reading CPU quota directly
+            if (try_read_cgroup_v2_num_cpus_from(mnt_dir, num_cpus))
+                return num_cpus; // Successfully read number of CPUs for cgroup v2
+
+            if (!*paths_cache.v2_relative_path)
+                cache_relative_path_for(cgroup_fd, paths_cache);
+
+            // Now try reading including relative path
+            if (std::snprintf(dir, PATH_MAX, "%s/%s", mnt_dir, paths_cache.v2_relative_path) >= 0)
+                try_read_cgroup_v2_num_cpus_from(dir, num_cpus);
+            return num_cpus;
+        }
+
+        __TBB_ASSERT(std::strncmp(mnt_type, "cgroup", 6) == 0, "Unexpected cgroup type");
+
+        // TODO: Before opening the path, make sure the mnt_opts entry contains "cpu" controller
+        if (try_read_cgroup_v1_num_cpus_from(mnt_dir, num_cpus))
+            return num_cpus; // Successfully read number of CPUs for cgroup v1
+
+        if (!*paths_cache.v1_relative_path)
+            cache_relative_path_for(cgroup_fd, paths_cache);
+
+        if (std::snprintf(dir, PATH_MAX, "%s/%s", mnt_dir, paths_cache.v1_relative_path) >= 0)
+            try_read_cgroup_v1_num_cpus_from(dir, num_cpus);
+        return num_cpus;
+    }
+
+    static bool is_cpu_restriction_possible(std::FILE* cgroup_fd, cgroup_paths& relative_paths_cache) {
+        cache_relative_path_for(cgroup_fd, relative_paths_cache);
+
+        bool has_cgroup_v1_cpu_controller = *relative_paths_cache.v1_relative_path;
+        bool has_cgroup_v2 = *relative_paths_cache.v2_relative_path;
+        bool is_cgroup_v2_with_root_path =
+            relative_paths_cache.v2_relative_path[0] == '/' && !relative_paths_cache.v2_relative_path[1];
+        if (!has_cgroup_v2 && !has_cgroup_v1_cpu_controller) {
+            return false; // No CPU constraints found
+        } else if (has_cgroup_v1_cpu_controller || !is_cgroup_v2_with_root_path) {
+            return true;
+        }
+        __TBB_ASSERT(!has_cgroup_v1_cpu_controller && has_cgroup_v2 && is_cgroup_v2_with_root_path, nullptr);
+
+        // Usually PID 1 is the init process
```

Comment:
Are there exceptions? :)

Reply:
Ah, you meant the hosting OS init process. Well, then perhaps "usually" is not the best word since containers are also usual nowadays. Anyway, expanding the comment a bit might be useful.